### ISSUE TITLE
Proposal to change default padding value from VALID to SAME, for convolutional APIs

### ISSUE THREAD

**Issue by Sayem-Mohammad-Imtiaz** [keras-team/keras#14410]
We believe that the value of the default ‘VALID’ padding should be changed to ‘SAME’ padding for convolutional APIs. This is because in majority of use cases the choice ‘VALID’ is not appropriate. This observation is based on an analysis of the latest snapshot of about 10K top-rated open source deep learning applications.

In our analysis, we found that a staggering amount of convolutional API (**keras and tf.keras**) usages (about 78%) end up using 'SAME' padding mode. This is even after considering all suspicious cases (where a variable has been used for instance instead of direct string value) in favor of 'VALID' padding.  For example, in project: **qqwweee/keras-yolo3/**, revision: **9ef8628f8583bc7494eeab1519f2ef2dd16fff0d**, file: **convert.py,** method: **_main**, padding in line 97 has been modified which has been used by **Conv2D** API in line **166**. Although, we can see that it can be SAME padding sometime, we treated such suspicious cases in favor of default VALID padding. The full Github URL for this: **https://github.com/qqwweee/keras-yolo3/blob/9ef8628f8583bc7494eeab1519f2ef2dd16fff0d/convert.py.**  So, in summary, about 78% of the time default VALID padding flips to SAME padding. 

We have reasoned that convolutional APIs are conventionally not meant to be down-sampler which is what VALID padding does as opposed to pooling layers. To further support our proposed change, we have found a set of **tensorflow** APIs and highly popular **tensorflow.contrib** convolutional APIs (the list of those APIs included below) that already use SAME padding as opposed to VALID padding. Interestingly, for them we found that, only 13% of such usages change. In other words, about 87% of all usages retain default value.

 
We also verified default VALID padding used by pooling APIs. Since pooling APIs are traditionally meant to be down-sampler, we can make an educated guess that they would be changed less often. Interestingly, we found that their flip rate is only 31% as opposed to be 78% for convolutional VALID APIs. Had it been SAME padding as default for pooling API, we could have observed a very different scenario.

Although it may look trivial, as an application developer, this is very important from the perspective of software evolution. A high change-rate for those APIs increases the potential to disrupt otherwise stable models. In our opinion, it is in all of our best interest to give less reasons for changes as changes are often expensive. Besides, in cases such as this where there are obvious trade-offs (when using valid, we lose some information but get model efficiency for reduced size) between choices, users are better off making such decisions for themselves.

**Default SAME padding Convolutional API:**
- tensorflow.nn.conv2d_transpose
- tensorflow.nn.conv3d_transpose
- tensorflow.nn.deconv2d
- tensorflow.compat.v1.nn.conv2d_transpose
- tensorflow.contrib.layers.conv2d
- tensorflow.contrib.layers.conv2d_transpose
- tensorflow.contrib.layers.separable_conv2d
- tensorflow.contrib.slim.conv2d
- tensorflow.contrib.slim.layers.conv2d
- tensorflow.contrib.slim.ops.conv2d
